pkg/rules/rules.go

```diff
@@ -0,0 +1,133 @@
+)
+
+type Options struct {
+	QueueCapacity   int
+	OutageTolerance time.Duration
+	ForGracePeriod  time.Duration
+	ResendDelay     time.Duration
+}
+
+type manager struct {
+	rulesManager     *prom_rules.Manager
+	notifierManager  *notifier.Manager
+	discoveryManager *discovery.Manager
+}
+
+func NewManager(ctx context.Context, r prometheus.Registerer, client *pgclient.Client, opts *Options) (*manager, error) {
+	discoveryManagerNotify := discovery.NewManager(ctx, log.GetLogger(), discovery.Name("notify"))
+
+	notifierManager := notifier.NewManager(&notifier.Options{
+		QueueCapacity: opts.QueueCapacity,
+		Registerer:    r,
+		Do:            do,
+	}, log.GetLogger())
+
+	// For the moment, we do not have any external UI url, hence we provide an empty one.
+	parsedUrl, err := url.Parse("")
+	if err != nil {
+		return nil, fmt.Errorf("parsing UI-URL: %w", err)
+	}
+
+	rulesManager := prom_rules.NewManager(&prom_rules.ManagerOptions{
+		Appendable:      adapters.NewIngestAdapter(client.Ingestor()),
+		Queryable:       adapters.NewQueryAdapter(client.Queryable()),
+		Context:         ctx,
+		ExternalURL:     parsedUrl,
+		Logger:          log.GetLogger(),
+		NotifyFunc:      sendAlerts(notifierManager, parsedUrl.String()),
+		QueryFunc:       engineQueryFunc(client.QueryEngine(), client.Queryable()),
+		Registerer:      r,
+		OutageTolerance: opts.OutageTolerance,
+		ForGracePeriod:  opts.ForGracePeriod,
+		ResendDelay:     opts.ResendDelay,
+	})
+	return &manager{
+		rulesManager:     rulesManager,
+		notifierManager:  notifierManager,
+		discoveryManager: discoveryManagerNotify,
+	}, nil
+}
+
+func (m *manager) ApplyConfig(cfg *prometheus_config.Config) error {
+	if err := m.applyDiscoveryManagerConfig(cfg); err != nil {
+		return err
+	}
+	if err := m.applyNotifierManagerConfig(cfg); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *manager) applyDiscoveryManagerConfig(cfg *prometheus_config.Config) error {
+	c := make(map[string]discovery.Configs)
+	for k, v := range cfg.AlertingConfig.AlertmanagerConfigs.ToMap() {
+		c[k] = v.ServiceDiscoveryConfigs
+	}
+	return errors.WithMessage(m.discoveryManager.ApplyConfig(c), "error applying config to discover manager")
+}
+
+func (m *manager) applyNotifierManagerConfig(cfg *prometheus_config.Config) error {
+	return errors.WithMessage(m.notifierManager.ApplyConfig(cfg), "error applying config to notifier manager")
+}
+
+func (m *manager) Update(interval time.Duration, files []string, externalLabels labels.Labels, externalURL string) error {
+	return errors.WithMessage(m.rulesManager.Update(interval, files, externalLabels, externalURL), "error updating the rules manager")
+}
+
+// Run runs the managers and blocks on either a graceful exit or on error.
+func (m *manager) Run() error {
+	var g run.Group
+
+	g.Add(func() error {
+		log.Info("msg", "starting discovery manager...")
+		return errors.WithMessage(m.discoveryManager.Run(), "error running discovery manager")
+	}, func(err error) {})
```

Comment:
shouldn't we log `log.Info("msg", "stopping discovery manager")` in the stop func?

Reply:
Sure will add that

Reply:
Well, I didn't implement that because the discovery manager does not have any `Stop()` function. I guess is it depends on the context of the notifier manager (which is used to create the discovery manager). So stopping the notifier manager should stop the discovery manager. But for log, we can add to make the lifecycle cleaner.